### PR TITLE
Create generic exception handler

### PR DESCRIPTION
**Pull request by ChernykhVyacheslav**
Created generic exception handlerf or CustomEntityNotFoundException and CustomFailedToDeleteEntityException

https://trello.com/c/65HgbOCw/145-create-generic-exception-handler

### CODE REVIEW

File: controller/src/main/java/com/softserve/borda/controllers/BoardController.java [hunk 1]
```diff
@@ -55,104 +53,72 @@ public Board getBoard(@PathVariable Long id) {
...
+        userBoardRelation.setBoard(board);
+        userBoardRelation.setUser((jwtConvertor.getUserByJWT(authorization)));
+
+        userBoardRelation.setBoardRole(userBoardRelationService
+                .getBoardRoleByName(BoardRole.BoardRoles.OWNER.name()));
+
+        board.setUserBoardRelations(Collections.singletonList(userBoardRelation));
+
+        return new ResponseEntity<>(modelMapper.map(
+                boardService.createOrUpdate(board),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardController.java [hunk 2]
```diff
@@ -55,104 +53,72 @@ public Board getBoard(@PathVariable Long id) {
...
-                    boardService.createOrUpdate(existingBoard),
-                    BoardFullDTO.class), HttpStatus.OK);
-        } catch (Exception e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        Board existingBoard = boardService.getBoardById(id);
+        BeanUtils.copyProperties(board, existingBoard);
+        return new ResponseEntity<>(modelMapper.map(
+                boardService.createOrUpdate(existingBoard),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardController.java [hunk 3]
```diff
@@ -55,104 +53,72 @@ public Board getBoard(@PathVariable Long id) {
...
-                            .stream().map(boardList -> modelMapper.map(boardList,
-                            BoardListDTO.class)).collect(Collectors.toList()),
-                    HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        return new ResponseEntity<>(
+                boardService.getAllBoardListsByBoardId(id)
+                        .stream().map(boardList -> modelMapper.map(boardList,
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardController.java [hunk 4]
```diff
@@ -55,104 +53,72 @@ public Board getBoard(@PathVariable Long id) {
...
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
-        }
+        BoardList boardList = new BoardList();
+        boardList.setName(boardListDTO.getName());
+        boardList = boardListService.createOrUpdate(boardList);
+        return new ResponseEntity<>(
+                modelMapper.map(
+                        boardService.addBoardListToBoard(
+                                boardService.getBoardById(boardId), boardList),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardListController.java [hunk 5]
```diff
@@ -34,107 +32,74 @@
...
-                            boardListService.getBoardListById(boardListId),
-                            BoardListDTO.class),
-                    HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        return new ResponseEntity<>(
+                modelMapper.map(
+                        boardListService.getBoardListById(boardListId),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardListController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardListController.java [hunk 6]
```diff
@@ -34,107 +32,74 @@
...
-                    HttpStatus.OK);
-        } catch (Exception e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        BoardList existingBoardList = boardListService.getBoardListById(boardListId);
+        BeanUtils.copyProperties(boardList, existingBoardList);
+        return new ResponseEntity<>(
+                modelMapper.map(
+                        boardListService.createOrUpdate(existingBoardList),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardListController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardListController.java [hunk 7]
```diff
@@ -34,107 +32,74 @@
...
-                    boardListService.getAllTicketsByBoardListId(boardListId)
-                            .stream().map(ticket -> modelMapper.map(ticket,
-                            TicketDTO.class)).collect(Collectors.toList()),
-                    HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        return new ResponseEntity<>(
+                boardListService.getAllTicketsByBoardListId(boardListId)
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardListController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardListController.java [hunk 8]
```diff
@@ -34,107 +32,74 @@
...
-                    HttpStatus.OK);
-        } catch (Exception e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
-        }
+        Ticket ticket = modelMapper.map(ticketDTO, Ticket.class);
+        ticket = ticketService.createOrUpdate(ticket);
+        return new ResponseEntity<>(
+                modelMapper.map(
+                        boardListService.addTicketToBoardList(
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardListController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardListController.java [hunk 9]
```diff
@@ -34,107 +32,74 @@
...
-        } catch (CustomFailedToDeleteEntityException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(modelMapper.map(
-                    boardListService.getBoardListById(boardListId),
-                    BoardListDTO.class),
-                    HttpStatus.NOT_FOUND);
-        }
+        return new ResponseEntity<>(
+                modelMapper.map(
+                        boardListService.deleteTicketFromBoardList(
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardListController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/BoardListController.java [hunk 10]
```diff
@@ -34,107 +32,74 @@
...
+                                                                     @PathVariable Long ticketId) {
+        BoardList oldBoardList = boardListService.getBoardListById(oldBoardListId);
+        BoardList newBoardList = boardListService.getBoardListById(newBoardListId);
+        Ticket ticket = ticketService.getTicketById(ticketId);
+        oldBoardList.getTickets().remove(ticket);
+        newBoardList.getTickets().add(ticket);
+        boardListService.createOrUpdate(oldBoardList);
+        boardListService.createOrUpdate(newBoardList);
+        return new ResponseEntity<>(
+                modelMapper.map(
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/BoardListController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 11]
```diff
@@ -30,135 +28,84 @@
...
-            return new ResponseEntity<>(modelMapper.map(
-                    userService.getUserById(id)
-                    ,
-                    UserSimpleDTO.class), HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        return new ResponseEntity<>(userService.getAll().stream()
+                .map(user -> modelMapper.map(user, UserSimpleDTO.class))
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java:35]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 12]
```diff
@@ -30,135 +28,84 @@
...
-            return new ResponseEntity<>(modelMapper.map(
-                    userService.createOrUpdate(user),
-                    UserSimpleDTO.class), HttpStatus.CREATED);
-        } catch (Exception e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
-        }
+        User user = modelMapper.map(userDTO, User.class);
+        return new ResponseEntity<>(modelMapper.map(
+                userService.createOrUpdate(user),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 13]
```diff
@@ -30,135 +28,84 @@
...
-                    jwtConvertor.getUserByJWT(id),
-                    UserSimpleDTO.class), HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+    @GetMapping
+    public ResponseEntity<UserSimpleDTO> getUserByToken(@RequestHeader String authorization) {
+        return new ResponseEntity<>(modelMapper.map(
+                jwtConvertor.getUserByJWT(authorization),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 14]
```diff
@@ -30,135 +28,84 @@
...
-                    userService.createOrUpdate(existingUser),
-                    UserSimpleDTO.class), HttpStatus.OK);
-        } catch (Exception e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+        User existingUser = userService.getUserById(id);
+        BeanUtils.copyProperties(user, existingUser);
+        return new ResponseEntity<>(modelMapper.map(
+                userService.createOrUpdate(existingUser),
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 15]
```diff
@@ -30,135 +28,84 @@
...
-
-                    HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+    @GetMapping("/boards")
+    public ResponseEntity<List<BoardFullDTO>> getBoardsByUser(@RequestHeader String authorization) {
+        return new ResponseEntity<>(userService.getBoardsByUser(
+                jwtConvertor.getUserByJWT(authorization))
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 16]
```diff
@@ -30,135 +28,84 @@
...
-                    HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+    @GetMapping("/boardsByRole/{boardRoleId}")
+    public ResponseEntity<List<BoardFullDTO>> getBoardsByUserAndBoardRoleId(@RequestHeader String authorization,
+                                                                            @PathVariable Long boardRoleId) {
+        return new ResponseEntity<>(
+                userService.getBoardsByUserAndBoardRoleId(
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Don't write logic in a single row.

File: controller/src/main/java/com/softserve/borda/controllers/UserController.java [hunk 17]
```diff
@@ -30,135 +28,84 @@
...
-                            .collect(Collectors.toList()),
-                    HttpStatus.OK);
-        } catch (CustomEntityNotFoundException e) {
-            log.severe(e.getMessage());
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
-        }
+    @GetMapping("/comments")
+    public ResponseEntity<List<CommentDTO>> getCommentsByUser(@RequestHeader String authorization) {
+        return new ResponseEntity<>(
+                userService.getAllCommentsByUser(jwtConvertor.getUserByJWT(authorization))
```
Comment (achehrynets) [inline controller/src/main/java/com/softserve/borda/controllers/UserController.java]: Don't write logic in a single row.

File: shared/src/main/java/com/softserve/borda/config/RestResponseEntityExceptionHandler.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+@Log
+public class RestResponseEntityExceptionHandler
+        extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(value
+            = {CustomEntityNotFoundException.class, CustomFailedToDeleteEntityException.class})
```
Comment (achehrynets) [inline shared/src/main/java/com/softserve/borda/config/RestResponseEntityExceptionHandler.java]: What if an entity cannot be removed due to some constraints? (e.g. it is a part of FR for some third entitty)

**Review by achehrynets — CHANGES_REQUESTED** [review]
Requested changes.

**Review by jules-n — APPROVED** [review]
Good job

**Review by HorbatenkoYehor — APPROVED** [review]
+1

File: controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java [hunk 19]
```diff
@@ -2,29 +2,29 @@
...
-    private final com.softserve.borda.config.jwt.JwtProvider jwtProvider;
+    private final JwtProvider jwtProvider;
 
     private final JWTUserRepository jwtUserRepository;
 
-
     public JwtConvertor(com.softserve.borda.config.jwt.JwtProvider jwtProvider, JWTUserRepository jwtUserRepository) {
         this.jwtProvider = jwtProvider;
         this.jwtUserRepository = jwtUserRepository;
     }
```
Comment (Dchad99) [inline controller/src/main/java/com/softserve/borda/config/jwt/JwtConvertor.java]: You may use @AllArgsConstructor

**Pull request merged**